### ISSUE TITLE
Add more NRBC treatment in the 1D Euler solver

### ISSUE THREAD

**Issue by yungyuc** [solvcon/solvcon#330]
PR #326 implemented a NRBC treatment for the 1D Euler solver.  There are other variants that we should try.

The associated issue for adding notes: https://github.com/solvcon/mmnote/issues/31

**Comment by Gene0315** [comment]
<img width="617" alt="截圖 2024-05-01 上午1 14 46" src="https://github.com/solvcon/modmesh/assets/165459695/6f6eb417-293d-49c0-abb9-aa14e98dce26">
In the CESE method, the points j on boundary have a lack of data from outside points, so it is difficult to predict their value at next time step. Maybe one possible approach is to extend the half dx cell point outside the boundary (red cross icon in the diagram). Then, these j points on boundary will can be solved. Alternatively, we can avoid the situation by excluding these points j when plotting.

**Comment by yungyuc** [comment]
@Gene0315 I agree that we can also make the plotting look nicer.  That can use a different issue to track.  Could you please create one?  (And associate that one with this.)

**Comment by Gene0315** [comment]
I try to use the second NRBC method with lambda=0. I think the zig-zag array could be shown simpely in the figure. The number 0~3 is the first index of m_so0 and m_so1.
<img width="552" alt="截圖 2024-05-28 中午12 05 13" src="https://github.com/solvcon/modmesh/assets/165459695/e1ef5d57-c393-417f-8c91-235ccd8af52c">

As I see it, the value at boundary point, such as m_so0(0, 1:3) and m_so1(0, 1:3), should be equal to the m_so0(1, 1:3) and m_so1(1, 1:3) before calculating the new value in next time step. Thus, I revised the “treat_boundary_so0()” and “treat_boundary_so1()” in “Euler1DCore.cp”.
<img width="787" alt="截圖 2024-05-28 中午12 13 20" src="https://github.com/solvcon/modmesh/assets/165459695/0f8195c4-3b96-4db2-aeb5-1deed5e84e19">
<img width="784" alt="截圖 2024-05-28 中午12 13 34" src="https://github.com/solvcon/modmesh/assets/165459695/279ee137-1ae9-4412-a413-da36895c13e5">

And then, I changed the sequence of “treat_boundary_so0”, “treat_boundary_so1”, and “march_half_so0”. The sections I changed including
 - wrap_onedim.cpp
<img width="855" alt="截圖 2024-05-28 中午12 16 18" src="https://github.com/solvcon/modmesh/assets/165459695/279582cf-e0fc-46a2-aca1-3fe74fdf8d3b">

 - core.hpp
<img width="847" alt="截圖 2024-05-28 中午12 24 05" src="https://github.com/solvcon/modmesh/assets/165459695/7b81d182-53ae-47af-9874-fbcd5af8ead2">

 - wrap_spacetime.cpp
<img width="846" alt="截圖 2024-05-28 中午12 25 10" src="https://github.com/solvcon/modmesh/assets/165459695/02d46899-647b-4fca-803c-bced292ce38b">

  - wrap_spacetime.hpp
<img width="830" alt="截圖 2024-05-28 中午12 29 47" src="https://github.com/solvcon/modmesh/assets/165459695/f4d87be5-25b6-41c0-888b-23cd4625a83e">

 - test_onedim_euler.py
<img width="829" alt="截圖 2024-05-28 中午12 32 46" src="https://github.com/solvcon/modmesh/assets/165459695/aab183bb-b11d-45e3-bcf5-27d9879e47e5">

  - test_spacetime_solution.py
<img width="827" alt="截圖 2024-05-28 中午12 32 38" src="https://github.com/solvcon/modmesh/assets/165459695/57f4b91b-d648-49f9-922e-2ce0f7365f6e">

But, after 200 time steps, the result showed 
<img width="1075" alt="截圖 2024-05-28 中午12 33 36" src="https://github.com/solvcon/modmesh/assets/165459695/6ad4d300-44a9-476c-a70c-82bd65b29dab">

**Comment by yungyuc** [comment]
> @Gene0315 I agree that we can also make the plotting look nicer. That can use a different issue to track. Could you please create one? (And associate that one with this.)

@Gene0315 I wonder what is your response to this request?

And please do not attach images for showing code or diff like in your comment https://github.com/solvcon/modmesh/issues/330#issuecomment-2134329422 .  When you want to show plain text, use [plain text format of markdown](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-code).  For example, you can quote code diff like:

```diff
diff --git a/cpp/modmesh/onedim/Euler1DCore.cpp b/cpp/modmesh/onedim/Euler1DCore.cpp
index 8065d79..2eb9bde 100644
--- a/cpp/modmesh/onedim/Euler1DCore.cpp
+++ b/cpp/modmesh/onedim/Euler1DCore.cpp
@@ -179,7 +179,7 @@ void Euler1DCore::treat_boundary_so1()
     // Set outside value from inside value.
     {
         // Left boundary.
-        size_t const ic = 1;
+        size_t const ic = 0;
         m_so1(ic, 0) = m_so1(ic + 1, 0);
         m_so1(ic, 1) = m_so1(ic + 1, 1);
         m_so1(ic, 2) = m_so1(ic + 1, 2);
```

Code pasted in an image cannot be searched and is useless.  We should not do it in GitHub.

It is easier to discuss for your question in https://github.com/solvcon/modmesh/issues/330#issuecomment-2134329422 in a pull request.  Please create one for it.
